File: server/src/models/Session.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface ISession extends Document {
  userId: mongoose.Types.ObjectId;
  refreshTokenHash: string;
  // The previous hash and when it was rotated out — kept briefly so that
  // multiple browser tabs racing to refresh the same (about-to-expire)
  // token don't trip reuse-detection on each other. See refreshToken().
  previousRefreshTokenHash?: string;
  previousTokenRotatedAt?: Date;
  userAgent?: string;
  ipAddress?: string;
  isValid: boolean;
  createdAt: Date;
  expiresAt: Date;
}

const SessionSchema = new Schema<ISession>({
  userId: { type: Schema.Types.ObjectId, ref: 'User', required: true, index: true },
  refreshTokenHash: { type: String, required: true, unique: true, index: true },
  previousRefreshTokenHash: { type: String },
  previousTokenRotatedAt: { type: Date },
  userAgent: { type: String },
  ipAddress: { type: String },
  isValid: { type: Boolean, default: true, required: true },
  createdAt: { type: Date, default: Date.now },
  expiresAt: { type: Date, required: true }
});

export const Session = mongoose.models.Session || mongoose.model<ISession>('Session', SessionSchema);

File: mobile/lib/models/session.dart
class UserSession {
  final String id;
  final String? userAgent;
  final String? ipAddress;
  final bool isValid;
  final DateTime? createdAt;
  final DateTime? expiresAt;

  UserSession({
    required this.id,
    this.userAgent,
    this.ipAddress,
    this.isValid = true,
    this.createdAt,
    this.expiresAt,
  });

  factory UserSession.fromJson(Map<String, dynamic> json) {
    return UserSession(
      id: (json['_id'] ?? json['id'] ?? '').toString(),
      userAgent: json['userAgent'],
      ipAddress: json['ipAddress'],
      isValid: json['isValid'] ?? true,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      expiresAt: json['expiresAt'] != null ? DateTime.parse(json['expiresAt']) : null,
    );
  }
}
